src/python/dxpy/dx_extract_utils/InputsValidator.py

```diff
@@ -0,0 +1,273 @@
+# TODO: check number of arguments and type manually
+# TODO: schema versioning handling?
+
+from __future__ import print_function
+
+
+class InputsValidator:
+    """
+    InputsValidator class for extract_assay expresion. Checks for invalid input combinations set by a JSON schema.
+
+    The schema is a dictionary with the following structure:
+
+    {
+    "schema_version": "1.0",
+    "parser_args":[
+                  "path"
+                  "...",
+                  ]
+    "condition_1": {
+        "properties": {
+            "items": ["path", "json_help"],
+        },
+        "condition": "at_least_one_required",
+        "error_message": {
+            "message": "..."
+        },
+    },
+    "condition_2": {
+        "properties": {
+            "main_key": "path",
+            "items": ["output","delim"]
+        },
+        "condition": "with_at_least_one_required",
+        "error_message": {
+            "message": "...",
+            "type": "warning"
+            },
+
+    """
+
+    def __init__(
+        self,
+        parser_dict,
+        schema,
+        error_handler=print,
+        warning_handler=print,
+    ):
+        self.parser_dict = parser_dict
+        self.schema = schema
+        self.error_handler = error_handler
+        self.warning_handler = warning_handler
+
+        self.schema_version = schema.get("schema_version")
+        self.conditions_funcs = {
+            "exclusive": "interpret_exclusive",
+            "exclusive_with_exceptions": "interpret_exclusive_with_exceptions",
+            "required": "interpret_required",
+            "at_least_one_required": "interpret_at_least_one_required",
+            "with_at_least_one_required": "interpret_with_at_least_one_required",
+            "with_none_of": "interpret_with_none_of",
+            "mutually_exclusive_group": "interpret_mutually_exclusive_group",
+        }
+
+    ### Schema methods ###
+    def validate_schema_conditions(self):
+        # Checking if all conditions exist
+        present_conditions = [
+            value.get("condition")
+            for key, value in self.schema.items()
+            if key != "schema_version" and key != "parser_args"
+        ]
+        not_found = set(present_conditions) - set(self.conditions_funcs)
+        if len(not_found) != 0:
+            self.error_handler("{} schema condition is not defined".format(not_found))
+
+    ### Checking general methods ###
+    def interpret_conditions(self):
+        for key, value in self.schema.items():
+            if key != "schema_version" and key != "parser_args":
+                condition = value.get("condition")
+                method_to_call = self.conditions_funcs.get(condition)
+                getattr(self, method_to_call)(key)
+
+    def throw_message(self, check):
+        type = self.schema.get(check).get("error_message").get("type")
```

Comment:
You can save the retrieved "check" object and pass that one down. Or better, extract the message into a variable also and pass just the message to the `throw_` method. That provides the lowest scope possible for the downstream method and in turn it's testing and usage becomes easier (str vs object).